### Opinion Review 데이터 세트를 이용한 문서 군집화 수행
- 비지도 학습이나, 여기서는 이해를 위해서 파일명을 답처럼 생각해서 진행

- https://archive.ics.uci.edu/ml/datasets/Opinosis+Opinion+%26frasl%3B+Review

In [1]:
# 파일 읽어서 데이터 프레임으로 변환
import numpy as np
import pandas as pd
import glob, os
# glob : 지정한 파일 목록 반환 모듈

In [3]:
# 리뷰의 파일 목록 구하기
# glob와 os.path.join function 을 이용하여 토픽 디렉토리안의 모든 리뷰 파일 가져오기
all_files = glob.glob(
      os.path.join('./data/uci_reviews/OpinosisDataset1.0/topics/'   # 파일 목록 가져올 위치 지정
    , '*.data')    # 가져올 파일의 패턴 지정
)
all_files

['./data/uci_reviews/OpinosisDataset1.0/topics\\accuracy_garmin_nuvi_255W_gps.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\bathroom_bestwestern_hotel_sfo.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\battery-life_amazon_kindle.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\battery-life_ipod_nano_8gb.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\battery-life_netbook_1005ha.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\buttons_amazon_kindle.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\comfort_honda_accord_2008.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\comfort_toyota_camry_2007.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\directions_garmin_nuvi_255W_gps.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\display_garmin_nuvi_255W_gps.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/topics\\eyesight-issues_amazon_kindle.txt.data',
 './data/uci_reviews/OpinosisDataset1.0/top

In [4]:
len(all_files)

51

In [5]:
all_files[0]

'./data/uci_reviews/OpinosisDataset1.0/topics\\accuracy_garmin_nuvi_255W_gps.txt.data'

In [8]:
# 연습 : 1개의 파일 데이터프레임으로 변환
df = pd.read_table(all_files[0], index_col=None, header=None, encoding='latin1')

In [30]:
df

0
0     The voice prompts and maps are wonderful esp...
1   I also thought the the voice prompts of the 75...
2    The 750 sounds more like a real female voice,...
3    I didn't have the problems with the sounds of...
4                   Some of the voices are annoying .
..                                                ...
84  Others have said that the voice directions lea...
85   I must say, though, that I believed it would ...
86      The voice commands are excellent and timely .
87   The model 255W that we received seemed to tak...
88   The voice is very clear and gives plenty of w...

[89 rows x 1 columns]

In [24]:
str(df)

"                                                    0\n0                      , and is very, very accurate .\n1    but for the most part, we find that the Garmi...\n2    This function is not accurate if you don't le...\n3    It provides immediate alternatives if the rou...\n4    I've used other GPS units, as well as GPS bui...\n..                                                ...\n62                      seems to be rather accurate .\n63   It was accurate on determing original directi...\n64                  Highly accurate, POIs are great .\n65   I can't believe how accurate and detailed the...\n66   Speed of calculation, accuracy, and simplicit...\n\n[67 rows x 1 columns]"

In [13]:
# 이 예제에서의 파일명 자체가 답처럼 사용될 수 있다.
all_files[0]

'./data/uci_reviews/OpinosisDataset1.0/topics\\accuracy_garmin_nuvi_255W_gps.txt.data'

In [19]:
# 파일명 분리
all_files[0].split('\\')[-1].split('.')[0]

'accuracy_garmin_nuvi_255W_gps'

##  전체 코드 완성

In [43]:
import numpy as np
import pandas as pd
import glob, os

all_files = glob.glob(
      os.path.join('./data/uci_reviews/OpinosisDataset1.0/topics/'   # 파일 목록 가져올 위치 지정
    , '*.data')    # 가져올 파일의 패턴 지정
)

# for 문 돌리기

# 파일 이름 저장공간 생성
filename_list = [] 

# 데이터 저장 공간 생성
opinion_text = []

# 파일이름을 이용해서, 데이터 프레임 생성
    
# 개별파일들의 파일내용은 DataFrame 로딩후, 다시 스트링으로 변환하여 opinion_text 리스트로 취합
# 파일 이름 전체 얻기
for file_ in all_files:
    df = pd.read_table(file_, index_col=None, header=0, encoding='latin1')
    opinion_text.append(df.to_string())
    filename = file_.split('\\')[-1].split('.')[0]
    filename_list.append(filename)

In [44]:
len(opinion_text), len(filename_list)

(51, 51)

In [45]:
filename_list[50]

'voice_garmin_nuvi_255W_gps'

In [46]:
opinion_text[50]

"                                                                                                                                                                                                 The voice prompts and maps are wonderful especially when driving after dark .\n0                                                                                                                                                                               I also thought the the voice prompts of the 750 where more pleasant sounding than the 255w's .\n1                                                                                                                                                                                           The 750 sounds more like a real female voice, where the 255w sounds more robotic .\n2                                                                                                                                                                                      

### df.to_string() 하면 데이터 프레임의 텍스트가 열순서정보(ex : \n64)까지 포함된 채로 텍스트로 반환된다

In [47]:
# 파일명 list와 파일내용 list 객체를 하나의 데이터 프레임으로 말아서 문서 저장하기
doc_df = pd.DataFrame({
    'filename' : filename_list
    ,'opinion_text' : opinion_text
})
doc_df.head()

filename  \
0   accuracy_garmin_nuvi_255W_gps   
1  bathroom_bestwestern_hotel_sfo   
2      battery-life_amazon_kindle   
3      battery-life_ipod_nano_8gb   
4     battery-life_netbook_1005ha   

                                        opinion_text  
0                                                ...  
1                                                ...  
2                                                ...  
3                                                ...  
4                                                ...

In [48]:
doc_df['opinion_text'][0]

"                                                                                                                                                                                                                                                               , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This function is not accurate if you don't leave it in battery mode say, when you stop at the Cracker Barrell for lunch and to play one of those trangle games with the tees .\n2                                                                                                                                         

#### TF-IDF 피처 벡터화, TfidfVectorizer에서 피처 벡터화 수행 시 Lemmatization을 적용하여 토큰화

In [49]:
# 토큰화 : 단어 분리
# 벡터화 : 수치화
import string
string.punctuation   # 특수문자를 문자열로 반환

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [52]:
# ASCII 값 출력 함수
ord('!'), chr(33)

(33, '!')

In [53]:
dict(
    (ord(punct),None) for punct in '!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'
)

{33: None,
 34: None,
 35: None,
 36: None,
 37: None,
 38: None,
 39: None,
 40: None,
 41: None,
 42: None,
 43: None,
 44: None,
 45: None,
 46: None,
 47: None,
 58: None,
 59: None,
 60: None,
 61: None,
 62: None,
 63: None,
 64: None,
 91: None,
 92: None,
 93: None,
 94: None,
 95: None,
 96: None,
 123: None,
 124: None,
 125: None,
 126: None}

### 표제어 추출 (Lemmatization)
- 표제어(Lemma)는 한글로는 '표제어' 또는 '기본 사전형 단어' 정도의 의미를 갖습니다. 표제어 추출은 단어들로부터 표제어를 찾아가는 과정입니다. 표제어 추출은 단어들이 다른 형태를 가지더라도, 그 뿌리 단어를 찾아가서 단어의 개수를 줄일 수 있는지 판단합니다. 예를 들어서 am, are, is는 서로 다른 스펠링이지만 그 뿌리 단어는 be라고 볼 수 있습니다. 이 때, 이 단어들의 표제어는 be라고 합니다.

- 표제어 추출을 하는 가장 섬세한 방법은 단어의 형태학적 파싱을 먼저 진행하는 것입니다. 형태소란 '의미를 가진 가장 작은 단위'를 뜻합니다. 그리고 형태학(morphology)이란, 형태소로부터 단어들을 만들어가는 학문을 뜻합니다.

- 형태소는 두 가지 종류가 있습니다. 각각 어간(stem)과 접사(affix)입니다.

1) 어간(stem)
: 단어의 의미를 담고 있는 단어의 핵심 부분.

2) 접사(affix)
: 단어에 추가적인 의미를 주는 부분.

In [54]:
from nltk.stem import WordNetLemmatizer
import nltk
import string

remove_punct_dict = dict((ord(punct), None) for punct in string.punctuation)

In [55]:
lemmar = WordNetLemmatizer()

In [57]:
# 토큰화 함수 작성
def LemTokens(tokens):
    return [lemmar.lemmatize(token) for token in tokens]

def LemNormalize(text):
    return LemTokens(nltk.word_tokenize(text.lower().translate(remove_punct_dict)))

In [56]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [58]:
# 토큰분리 객체 생성
tfidf_vect = TfidfVectorizer(tokenizer=LemNormalize  # 위에서 만든 토큰화 함수
                             , stop_words='english'
                             , ngram_range=(1,2), min_df=0.05, max_df=0.85 )


In [59]:
#opinion_text 컬럼값으로 feature vectorization 수행
# feature_vect : 군집화할 학습데이터가 됨.
feature_vect = tfidf_vect.fit_transform(doc_df['opinion_text'])

C:\Users\TJ\anaconda3\envs\ml01-02\lib\site-packages\sklearn\feature_extraction\text.py:386: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['ha', 'le', 'u', 'wa'] not in stop_words.
  'stop_words.' % sorted(inconsistent))


In [62]:
feature_vect

<51x4611 sparse matrix of type '<class 'numpy.float64'>'
	with 30124 stored elements in Compressed Sparse Row format>

## 벡터화 된 피처로 5개의 군집으로 K-Means 군집화

In [63]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화 수행. 예제를 위해 동일한 클러스터링 결과 도출용 random_state=0 
km_cluster = KMeans(n_clusters=5, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_

# cluster_centers_ : 군집별 핵심 단어 추출
cluster_centers = km_cluster.cluster_centers_

In [67]:
doc_df['cluster_label'] = cluster_label
doc_df.head()

filename  \
0   accuracy_garmin_nuvi_255W_gps   
1  bathroom_bestwestern_hotel_sfo   
2      battery-life_amazon_kindle   
3      battery-life_ipod_nano_8gb   
4     battery-life_netbook_1005ha   

                                        opinion_text  cluster_label  
0                                                ...              0  
1                                                ...              1  
2                                                ...              3  
3                                                ...              3  
4                                                ...              3

In [72]:
# cluster_label ==0 추출, 파일명으로 정렬
doc_df[doc_df['cluster_label']==0].sort_values(by='filename')

filename  \
0     accuracy_garmin_nuvi_255W_gps   
8   directions_garmin_nuvi_255W_gps   
9      display_garmin_nuvi_255W_gps   
33   satellite_garmin_nuvi_255W_gps   
34      screen_garmin_nuvi_255W_gps   
43       speed_garmin_nuvi_255W_gps   
47   transmission_toyota_camry_2007   
48     updates_garmin_nuvi_255W_gps   

                                         opinion_text  cluster_label  
0                                                 ...              0  
8                                                 ...              0  
9                                                 ...              0  
33                                                ...              0  
34                                                ...              0  
43                                                ...              0  
47                                                ...              0  
48                                                ...              0

In [73]:
doc_df[doc_df['cluster_label']==1].sort_values(by='filename')

filename  \
1    bathroom_bestwestern_hotel_sfo   
13          food_holiday_inn_london   
14           food_swissotel_chicago   
15       free_bestwestern_hotel_sfo   
20   location_bestwestern_hotel_sfo   
21      location_holiday_inn_london   
24    parking_bestwestern_hotel_sfo   
28         price_holiday_inn_london   
32          room_holiday_inn_london   
30      rooms_bestwestern_hotel_sfo   
31          rooms_swissotel_chicago   
38    service_bestwestern_hotel_sfo   
39       service_holiday_inn_london   
40  service_swissotel_hotel_chicago   
45      staff_bestwestern_hotel_sfo   
46          staff_swissotel_chicago   

                                         opinion_text  cluster_label  
1                                                 ...              1  
13                                                ...              1  
14                                                ...              1  
15                                                ...              1  
20                                                ...              1  
21                                                ...              1  
24                                                ...              1  
28                                                ...              1  
32                                                ...              1  
30                                                ...              1  
31                                                ...              1  
38                                                ...              1  
39                                                ...              1  
40                                                ...              1  
45                                                ...              1  
46                                                ...              1

In [74]:
doc_df[doc_df['cluster_label']==2].sort_values(by='filename')

filename  \
6       comfort_honda_accord_2008   
7       comfort_toyota_camry_2007   
16  gas_mileage_toyota_camry_2007   
17     interior_honda_accord_2008   
18     interior_toyota_camry_2007   
22      mileage_honda_accord_2008   
25  performance_honda_accord_2008   
29      quality_toyota_camry_2007   
37        seats_honda_accord_2008   

                                         opinion_text  cluster_label  
6                                                 ...              2  
7                                                 ...              2  
16                                                ...              2  
17                                                ...              2  
18                                                ...              2  
22                                                ...              2  
25                                                ...              2  
29                                                ...              2  
37                                                ...              2

In [75]:
doc_df[doc_df['cluster_label']==3].sort_values(by='filename')

filename  \
2    battery-life_amazon_kindle   
3    battery-life_ipod_nano_8gb   
4   battery-life_netbook_1005ha   
26   performance_netbook_1005ha   
42          sound_ipod_nano_8gb   
49          video_ipod_nano_8gb   

                                         opinion_text  cluster_label  
2                                                 ...              3  
3                                                 ...              3  
4                                                 ...              3  
26                                                ...              3  
42      headphone jack i got a clear case for it a...              3  
49                                                ...              3

In [76]:
doc_df[doc_df['cluster_label']==4].sort_values(by='filename')

filename  \
5           buttons_amazon_kindle   
10  eyesight-issues_amazon_kindle   
11              features_windows7   
12            fonts_amazon_kindle   
19        keyboard_netbook_1005ha   
23       navigation_amazon_kindle   
27            price_amazon_kindle   
35           screen_ipod_nano_8gb   
36          screen_netbook_1005ha   
41       size_asus_netbook_1005ha   
44                 speed_windows7   
50     voice_garmin_nuvi_255W_gps   

                                         opinion_text  cluster_label  
5                                                 ...              4  
10                                                ...              4  
11                                                ...              4  
12                                                ...              4  
19                                                ...              4  
23                                                ...              4  
27                                                ...              4  
35                                                ...              4  
36                                                ...              4  
41                                                ...              4  
44                                                ...              4  
50                                                ...              4

- 꽤 잘 분류되었다

- 3개의 집합으로 해보자

In [101]:
from sklearn.cluster import KMeans

# 3개의 집합으로 군집화 
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_


# 소속 클러스터를 cluster_label 컬럼으로 할당하고 cluster_label 값으로 정렬
doc_df['cluster_label'] = cluster_label
doc_df.sort_values(by='cluster_label')

filename  \
0     accuracy_garmin_nuvi_255W_gps   
48     updates_garmin_nuvi_255W_gps   
44                   speed_windows7   
43       speed_garmin_nuvi_255W_gps   
42              sound_ipod_nano_8gb   
41         size_asus_netbook_1005ha   
36            screen_netbook_1005ha   
35             screen_ipod_nano_8gb   
34      screen_garmin_nuvi_255W_gps   
33   satellite_garmin_nuvi_255W_gps   
27              price_amazon_kindle   
26       performance_netbook_1005ha   
49              video_ipod_nano_8gb   
23         navigation_amazon_kindle   
19          keyboard_netbook_1005ha   
50       voice_garmin_nuvi_255W_gps   
9      display_garmin_nuvi_255W_gps   
4       battery-life_netbook_1005ha   
3        battery-life_ipod_nano_8gb   
2        battery-life_amazon_kindle   
8   directions_garmin_nuvi_255W_gps   
10    eyesight-issues_amazon_kindle   
11                features_windows7   
12              fonts_amazon_kindle   
5             buttons_amazon_kindle   
13          food_holiday_inn_london   
39       service_holiday_inn_london   
38    service_bestwestern_hotel_sfo   
1    bathroom_bestwestern_hotel_sfo   
14           food_swissotel_chicago   
20   location_bestwestern_hotel_sfo   
24    parking_bestwestern_hotel_sfo   
15       free_bestwestern_hotel_sfo   
31          rooms_swissotel_chicago   
30      rooms_bestwestern_hotel_sfo   
45      staff_bestwestern_hotel_sfo   
40  service_swissotel_hotel_chicago   
21      location_holiday_inn_london   
46          staff_swissotel_chicago   
32          room_holiday_inn_london   
28         price_holiday_inn_london   
47   transmission_toyota_camry_2007   
16    gas_mileage_toyota_camry_2007   
6         comfort_honda_accord_2008   
7         comfort_toyota_camry_2007   
29        quality_toyota_camry_2007   
22        mileage_honda_accord_2008   
18       interior_toyota_camry_2007   
17       interior_honda_accord_2008   
37          seats_honda_accord_2008   
25    performance_honda_accord_2008   

                                         opinion_text  cluster_label  
0                                                 ...              0  
48                                                ...              0  
44                                                ...              0  
43                                                ...              0  
42      headphone jack i got a clear case for it a...              0  
41                                                ...              0  
36                                                ...              0  
35                                                ...              0  
34                                                ...              0  
33                                                ...              0  
27                                                ...              0  
26                                                ...              0  
49                                                ...              0  
23                                                ...              0  
19                                                ...              0  
50                                                ...              0  
9                                                 ...              0  
4                                                 ...              0  
3                                                 ...              0  
2                                                 ...              0  
8                                                 ...              0  
10                                                ...              0  
11                                                ...              0  
12                                                ...              0  
5                                                 ...              0  
13                                                ...              1  
39                                                ...              1  
38        

### 군집별 핵심 단어 추출하기

In [102]:
# 단어 개수 확인하기
feature_vect.shape

(51, 4611)

- 단어 개수 : 4611개

- K-Means 객체의 cluster_centers속성은 개별 피처들의 클러스터 중심과의 상대 위치를 정규화된 숫자값으로 표시
- 0~1 까지의 값으로 표현되며 1에 가까울 수록 중심에 더 가깝다는 의미

In [103]:
cluster_centers = km_cluster.cluster_centers_
cluster_centers.shape

(3, 4611)

- 3개의 중심점 대비 4611개의 단어의 상대 위치가 정규화된 숫자값(0~1)으로 표시됨
- 즉 1에 가까운 숫자를 갖는 단어가 중심단어에 가깝다

In [104]:
cluster_centers

array([[0.01005322, 0.        , 0.        , ..., 0.00706287, 0.        ,
        0.        ],
       [0.        , 0.00099499, 0.00174637, ..., 0.        , 0.00183397,
        0.00144581],
       [0.        , 0.00092551, 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [105]:
cluster_centers[0,:]

array([0.01005322, 0.        , 0.        , ..., 0.00706287, 0.        ,
       0.        ])

In [110]:
# 큰수로 정렬한 후 인덱스 반환
cluster_centers.argsort()[::-1]

array([[   0, 2789, 2791, ..., 2706, 3612, 2252],
       [   0, 2275, 2272, ..., 3648, 2095, 3466],
       [4610, 3647, 1750, ..., 2342,  685, 3572]], dtype=int64)

In [114]:
# 군집별 top n 핵심단어
# 그 단어의 중심위치 상대값, 대상 파일명들을 반환함.
# clusters_num : 군집 개수
# feature_names : 분류되는 단어명
# cluster_data : 위에서 만든 df형태 데이터
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features=10):
    # return variable
    cluster_details = {}
    
    # cluster_centers array 의 값이 큰 순으로 정렬된 index 값을 반환
    # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함.  
    centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:,::-1]
    
    #개별 군집별로 iteration하면서 핵심단어, 그 단어의 중심 위치 상대값, 대상 파일명 입력
    for cluster_num in range(clusters_num):
        # 개별 군집별 정보를 담을 데이터 초기화. 
        cluster_details[cluster_num] = {}
        cluster_details[cluster_num]['cluster'] = cluster_num
        
        # cluster_centers_.argsort()[:,::-1] 로 구한 index 를 이용하여 top n 피처 단어를 구함. 
        top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
        top_features = [ feature_names[ind] for ind in top_feature_indexes ]
        
        # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함 
        top_feature_values = cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()
        
        # cluster_details 딕셔너리 객체에 개별 군집별 핵심 단어와 중심위치 상대값, 그리고 해당 파일명 입력
        # 딕셔너리에 벨류로 넣어야 하므로 tolist 로 리스트형태로 바꿔줌
        cluster_details[cluster_num]['top_features'] = top_features
        cluster_details[cluster_num]['top_features_value'] = top_feature_values
        filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
        filenames = filenames.values.tolist()
        cluster_details[cluster_num]['filenames'] = filenames
        
    return cluster_details

### 클러스터별 top feature 들의 단어와 파일명 출력

In [115]:
def print_cluster_details(cluster_details):
    for cluster_num, cluster_detail in cluster_details.items():   
        print('####### Cluster {0}'.format(cluster_num))
        print('Top features:', cluster_detail['top_features'])
        print('Reviews 파일명 :',cluster_detail['filenames'][:7])
        print('==================================================')

In [116]:
# 단어명이 들어가 있는 모델 객체 tfidf_vect 에서 feature_names() 추출
feature_names = tfidf_vect.get_feature_names()

In [117]:
cluster_details = get_cluster_details(km_cluster, doc_df, feature_names ,3, top_n_features=10)
print_cluster_details(cluster_details)

####### Cluster 0
Top features: ['screen', 'battery', 'keyboard', 'battery life', 'life', 'kindle', 'direction', 'video', 'size', 'voice']
Reviews 파일명 : ['accuracy_garmin_nuvi_255W_gps', 'battery-life_amazon_kindle', 'battery-life_ipod_nano_8gb', 'battery-life_netbook_1005ha', 'buttons_amazon_kindle', 'directions_garmin_nuvi_255W_gps', 'display_garmin_nuvi_255W_gps']
####### Cluster 1
Top features: ['room', 'hotel', 'service', 'staff', 'food', 'location', 'bathroom', 'clean', 'price', 'parking']
Reviews 파일명 : ['bathroom_bestwestern_hotel_sfo', 'food_holiday_inn_london', 'food_swissotel_chicago', 'free_bestwestern_hotel_sfo', 'location_bestwestern_hotel_sfo', 'location_holiday_inn_london', 'parking_bestwestern_hotel_sfo']
####### Cluster 2
Top features: ['interior', 'seat', 'mileage', 'comfortable', 'gas', 'gas mileage', 'transmission', 'car', 'performance', 'quality']
Reviews 파일명 : ['comfort_honda_accord_2008', 'comfort_toyota_camry_2007', 'gas_mileage_toyota_camry_2007', 'interior_hon

- 반환되는 cluster details 는 이중 딕셔너리 구조로 되어있으며, 군집화 레이블이 첫번째 key, 그리고 그 안에서 각 디테일 항목이 두번째 key가 되어 값들이 저장된다